llvm/lib/CodeGen/TargetLoweringObjectFileImpl.cpp

```diff
@@ -875,9 +875,12 @@ static MCSectionRepo *selectRepoSectionForGlobal(MCContext &Ctx,
     // Add new created digest to the TicketNodes.
     assert(!GO->getMetadata(LLVMContext::MD_repo_ticket) &&
            "TicketNode should be NULL!");
+    GlobalValue::LinkageTypes Linkage = GO->getLinkage();
+    if (Linkage == GlobalValue::PrivateLinkage)
+      Linkage = GlobalValue::InternalLinkage;
```

Comment:
> 
> 
> I can see a couple of problems with this:
> 
>     * We’re essentially duplicating the behavior of `RepoObjectWriter::toPstoreLinkage()`. That function knows that it can map GlobalValue PrivateLinkage and InternalLinkage to the repo `internal` linkage. I’m wary of (essentially) duplicating that code here.
> 
>     * `internal` linkage isn’t the only repo linkage with “n to 1” mapping from more than one LLVM linkage value: the same applies to repo `linkonce` linkage. This code doesn’t consider these cases.
>

Thanks for pointing out these problems. You are right.

Reply:
> 
> 
> One option might be to replicate all of the LLVM linkage types in the repo rather than reducing them in `toPstoreLinkage()`.
>

Good advice. I have implemented based on your suggestion. The changes affect the pstore, repo2obj and rld as well.

pstore: https://github.com/SNSystems/pstore/pull/42
rld: https://github.com/SNSystems/rld/pull/2